### PR TITLE
Moregenic

### PR DESCRIPTION
**Pull request by sbird**
A few more MP-GenIC fixes split out from scaledepgrowth.

f16eaac we decided to change after discussion with Andreu. 
d18b81b is a simple threading on MP-GenIC, and 6258764 adjusts some of the memory management in MP-GenIC to make it a bit clearer.

### CODE REVIEW

**Comment by sbird** [comment]
Yu, can you do me a quick review? Thanks.

File: libgenic/zeldovich.c [hunk 1]
```diff
@@ -158,15 +159,20 @@ void displacement_fields(int Type) {
...
     petapm_force_c2r(rho_k, regions, functions);
+    myfree(rho_k);
+    myfree(regions);
     petapm_force_finish();
     double maxdisp = 0;
     int i;
-    for(i = 0; i < NumPart; i ++) {
+    #pragma omp parallel for reduction(max:maxdisp)
+    for(i = 0; i < NumPart; i++)
+    {
```
Comment (rainwoodman) [inline libgenic/zeldovich.c]: Could you by the way make the `MPI_Reduce` a few lines lower in place? these one time variables are really eyesores.

File: libgenic/zeldovich.c [hunk 2]
```diff
@@ -301,7 +307,6 @@ typedef struct {
         ptrdiff_t total;
     } ORegion;
     int Nmesh[3];
-    size_t allocsize;
```
[the remark this answers is not in the document]
Comment (sbird) [reply libgenic/zeldovich.c]: Notice that fftsize is only used in petapm.c - this was previously only used for the memset. Once the memset is moved into petapm.c there is no need to store the total size at all

File: libgenic/zeldovich.c (hunk 1 above)
Comment (sbird) [reply libgenic/zeldovich.c]: I'll pull forward another part of the cleanup from the scaledepgrowth branch.

Comment (sbird) [reply libgenic/zeldovich.c]: How do you like it now?

File: libgenic/zeldovich.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbird) [reply libgenic/zeldovich.c]: I too would prefer to store allocsize in a local struct, but there isn't one handy for the gravpm user of petapm, unfortunately.

[the remark this answers is not in the document]
Comment (rainwoodman) [reply libgenic/zeldovich.c]: OK then.

File: libgenic/power.c [hunk 3]
```diff
@@ -204,17 +205,18 @@ double PowerSpec_Tabulated(double k, int Type)
...
   double trans = 1;
   /*Transfer table stores (T_type(k) / T_tot(k))^2*/
-  if(Type >= 0 && Type < MAXCOLS && transfer_table.Nentry > 0)
+  if(Type >= 0 && Type < MAXCOLS && transfer_table.Nentry > 0) {
       trans = gsl_interp_eval(transfer_table.mat_intp[Type], transfer_table.logk, transfer_table.logD[Type], logk, transfer_table.mat_intp_acc[Type]);
+  }
 
   double power = pow(10.0, logD) * trans;
 
   return power;
```
Comment (rainwoodman) [inline libgenic/power.c]: power or delta?

File: libgenic/zeldovich.c [hunk 4]
```diff
@@ -153,54 +154,52 @@ void displacement_fields(int Type) {
...
-    {
-        int k;
-        for(k = 0; k < 3; k++)
-        {
             ICP[i].Pos[k] += ICP[i].Vel[k];
             ICP[i].Vel[k] *= vel_prefac;
             ICP[i].Pos[k] = periodic_wrap(ICP[i].Pos[k]);
         }
     }
+    MPI_Reduce(&maxdisp, &maxdispall, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
```
Comment (rainwoodman) [inline libgenic/zeldovich.c]: I meant MPI_Allreduce(MPI_IN_PLACE, &maxdisp, 1, .....)

Comment (sbird) [reply libgenic/zeldovich.c]: Oh, sorry. Fixed.

File: libgenic/power.c (hunk 3 above)
Comment (sbird) [reply libgenic/power.c]: Fixed

**Comment by sbird** [comment]
You notice I also fixed the build system petaio rule from before: you were right that it was a hack. You can still do incremental builds because it doesn't check the git revision except when the build files changes.

**Comment by sbird** [comment]
Ok to merge? Thanks for the quick reviews, by the way.

**Pull request merged**